### PR TITLE
fix: move expensive client state iteration to tx

### PR DESCRIPTION
**Pull request by danwt**
## Description

----

Closes #XXX

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

----

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: proto/dymensionxyz/dymension/lightclient/tx.proto [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+package dymensionxyz.dymension.lightclient;
+
+import "cosmos/msg/v1/msg.proto";
+import "gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+import "google/protobuf/any.proto";
+
+option go_package = "github.com/dymensionxyz/dymension/v3/x/lightclient/types";
+
+service Msg {
```
Comment (VictorTrustyDev) [inline proto/dymensionxyz/dymension/lightclient/tx.proto:11]: maybe add a comment say this proto file was cloned to somewhere else so we won't forget

File: x/lightclient/ante/ibc_msgs_test.go [hunk 2]
```diff
@@ -17,6 +17,11 @@ type MockRollappKeeper struct {
 	stateInfos map[string]map[uint64]rollapptypes.StateInfo
 }
 
+func (m *MockRollappKeeper) GetLatestHeight(ctx sdk.Context, rollappId string) (uint64, bool) {
+	// TODO implement me
```
Comment (VictorTrustyDev) [inline x/lightclient/ante/ibc_msgs_test.go:21]: I guess no need this line

Comment (danwt) [reply x/lightclient/ante/ibc_msgs_test.go:21]: it's autogenerated, gonna dismiss as nit, but good point 👍

**Review by omritoptix — APPROVED** [review]
wrong godoc as we don't validate the consensus state here anymore:

https://github.com/dymensionxyz/dymension/pull/1575/files#diff-ba4c5cafcc04b485b66198c72aa462b98bfc4d80ffda3be72b94daf83cc9570bL30-L32

File: x/rollapp/keeper/latest_state_info_index.go [hunk 3]
```diff
@@ -36,6 +36,17 @@ func (k Keeper) GetLatestStateInfoIndex(
...
 
+func (k Keeper) GetLatestHeight(
+	ctx sdk.Context,
+	rollappId string,
+) (uint64, bool) {
+	info, ok := k.GetLatestStateInfo(ctx, rollappId)
+	if !ok {
+		return 0, false
+	}
+	return info.GetLatestHeight(), true
```
Comment (VictorTrustyDev) [inline x/rollapp/keeper/latest_state_info_index.go:47]: is this guaranteed latest height > 0?

**Review by VictorTrustyDev — APPROVED** [review]
LGTM

File: x/rollapp/keeper/latest_state_info_index.go (hunk 3 above)
Comment (danwt) [reply x/rollapp/keeper/latest_state_info_index.go:47]: why does it matter? If there is a state info, then the rollap definitely updated the hub

File: x/lightclient/keeper/canonical_client.go [hunk 4]
```diff
@@ -67,35 +93,41 @@ func (k Keeper) expectedClient() ibctm.ClientState {
 	return types.DefaultExpectedCanonicalClientParams()
 }
 
-var errChainIDMismatch = errors.New("chain id mismatch")
+var (
+	ErrNoMatch        = gerrc.ErrFailedPrecondition.Wrap("not at least one cons state matches the rollapp state")
+	ErrMismatch       = gerrc.ErrInvalidArgument.Wrap("consensus state mismatch")
+	ErrParamsMismatch = gerrc.ErrInvalidArgument.Wrap("params")
+)
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/keeper/canonical_client.go:100]: The types/errors pattern is not good IMO, it's not idiomatic to have one use errors defined far from the callsite

File: x/rollapp/keeper/latest_state_info_index.go (hunk 3 above)
Comment (VictorTrustyDev) [reply x/rollapp/keeper/latest_state_info_index.go:47]: then it's good

**Pull request merged**